### PR TITLE
Fix race condition in L2/ARP announcer

### LINKED ISSUE #681: Race condition
https://github.com/metallb/metallb/blob/main/internal/layer2/announcer.go#L158
There is a race condition.

We have a big cluster with many pod creation/deletion operations, so interfaces list on a node changes very often.
In our cluster MetalLB sometimes doesn't announce IPs (L2/ARP) because, at the moment when `gratuitous` function executes, one of the interfaces in `a.arps` array becomes dead and this loop crashes with an error, so no more gratuitous announces won't send.

I think we should ignore these errors and don't crash a loop.

### PR DESCRIPTION
**Pull request by maxpain**
Fixed #681

### CODE REVIEW

**Comment by champtar** [comment]
It's not obvious you are fixing a race condition, can you explain a bit ?

**Comment by maxpain** [comment]
@champtar 

We have a big cluster with many pod creation/deletion operations, so the interfaces list on a node often changes.
In our cluster, MetalLB sometimes doesn't announce IPs (L2/ARP) because, at the moment when a gratuitous function is executing, one of the interfaces in the `a.arps` array becomes dead, and this loop crashes with an error, so no more gratuitous announces on other interfaces will be sent.

Let's ignore these errors and don't crash the loop in case of an error.

**Comment by fedepaol** [comment]
@maxpain mind squashing the commits and extending the commit message with the rationale you just described?

File: internal/layer2/announcer.go [hunk 1]
```diff
@@ -196,13 +196,13 @@ func (a *Announce) gratuitous(ip net.IP) error {
 	if ip.To4() != nil {
 		for _, client := range a.arps {
 			if err := client.Gratuitous(ip); err != nil {
-				return err
+				level.Error(a.logger).Log("gratuitousAnnounce", "error", err, "ip", ip, "msg", "failed to make gratuitous IP announcement")
```
Comment (fedepaol) [inline internal/layer2/announcer.go:199]: Would it make sense to return here?

File: internal/layer2/announcer.go [hunk 2]
```diff
@@ -196,13 +196,13 @@ func (a *Announce) gratuitous(ip net.IP) error {
...
 			}
 		}
 	} else {
 		for _, client := range a.ndps {
 			if err := client.Gratuitous(ip); err != nil {
-				return err
+				level.Error(a.logger).Log("gratuitousAnnounce", "error", err, "ip", ip, "msg", "failed to make gratuitous IP announcement")
```
Comment (fedepaol) [inline internal/layer2/announcer.go:205]: same

File: internal/layer2/announcer.go (hunk 1 above)
Comment (fedepaol) [reply internal/layer2/announcer.go:199]: I meant return nil

Comment (russellb) [reply internal/layer2/announcer.go:199]: but then it would skip sending the GARP on all other interfaces?

Comment (russellb) [reply internal/layer2/announcer.go:199]: we should probably pick up discussion on #974

**Pull request closed without merging**